packages/ckeditor5-widget/src/widgettypearound/widgettypearound.js

```diff
@@ -0,0 +1,250 @@
+ */
+
+import Plugin from '@ckeditor/ckeditor5-core/src/plugin';
+import Paragraph from '@ckeditor/ckeditor5-paragraph/src/paragraph';
+import Template from '@ckeditor/ckeditor5-ui/src/template';
+
+import {
+	isTypeAroundWidget,
+	getWidgetTypeAroundPositions,
+	getClosestTypeAroundDomButton,
+	getTypeAroundButtonPosition,
+	getClosestWidgetViewElement
+} from './utils';
+
+import returnIcon from '../../theme/icons/return-arrow.svg';
+import '../../theme/widgettypearound.css';
+
+const POSSIBLE_INSERTION_POSITIONS = [ 'before', 'after' ];
+let CACHED_RETURN_ARROW_ICON;
+
+/**
+ * A plugin which allows users to type around widgets where normally it is impossible to place the caret due
+ * to limitations of web browsers. These "tight spots" occur, for instance, before (or after) a widget being
+ * the first (or last) child of its parent or between two block widgets.
+ *
+ * This plugin extends the {@link module:widget/widget~Widget `Widget`} plugin and inject a user interface
+ * with two buttons into each widget instance in the editor. Each of the buttons can be clicked by the
+ * user if the widget is next to the "tight spot". Once clicked, a paragraph is created with the selection anchored
+ * in it so that users can type (or insert content, paste, etc.) straight away.
+ *
+ * @extends module:core/plugin~Plugin
+ * @private
+ */
+export default class WidgetTypeAround extends Plugin {
+	/**
+	 * @inheritDoc
+	 */
+	static get requires() {
+		return [ Paragraph ];
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	static get pluginName() {
+		return 'WidgetTypeAround';
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	init() {
+		this._enableTypeAroundUIInjection();
+		this._enableDetectionOfTypeAroundWidgets();
+		this._enableInsertingParagraphsOnButtonClick();
+	}
+
+	/**
+	 * Inserts a new paragraph next to a widget element with the selection anchored in it.
+	 *
+	 * **Note**: This method is heavily user-oriented and will both focus the editing view and scroll
+	 * the viewport to the selection in the inserted paragraph.
+	 *
+	 * @private
+	 * @param {module:engine/view/element~Element} widgetViewElement The view widget element next to which a paragraph is inserted.
+	 * @param {String} position The position where the paragraph is inserted. Either `'before'` or `'after'` the widget.
+	 */
+	_insertParagraph( widgetViewElement, position ) {
+		const editor = this.editor;
+		const editingView = editor.editing.view;
+		let viewPosition;
+
+		if ( position === 'before' ) {
+			viewPosition = editingView.createPositionBefore( widgetViewElement );
+		} else {
+			viewPosition = editingView.createPositionAfter( widgetViewElement );
+		}
+
+		const modelPosition = editor.editing.mapper.toModelPosition( viewPosition );
+
+		editor.model.change( writer => {
+			const paragraph = writer.createElement( 'paragraph' );
+
+			writer.insert( paragraph, modelPosition );
+			writer.setSelection( paragraph, 0 );
+		} );
+
+		editingView.focus();
+		editingView.scrollToTheSelection();
```

Comment:
Shouldn't this be some generic editor feature that view would follow caret and not only widget specific? Something like "Caret movement" in WebStorm:

![Zrzut ekranu 2020-05-12 o 20 06 22](https://user-images.githubusercontent.com/1232187/81729423-385a5d80-948c-11ea-82c3-11710a7d06f0.png)